File: src/LogicExpression/Operands.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace LogicExpression
{
    public struct Operand : IComparable<Operand>
    {
        public char Symbol { get; }
        public bool Value { get; set; }

        public Operand(char symbol)
        {
            Symbol = symbol;
            Value = false;
        }

        public int CompareTo(Operand other)
        {
            return Symbol.CompareTo(other.Symbol);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
        public static bool IsValid(char c)
        {
            return c >= 'a' && c <= 'z';
        }
    }

    public class Operands : IEnumerable<char>
    {
        private const int OPCOUNT = 26;

        private readonly bool?[] _operands = new bool?[OPCOUNT];
        private int? _combinationCount = null;

        public int Count { get; private set; }

        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
        public int GetCombinations()
        {
            if (_combinationCount == null)
            {
                _combinationCount = (int)Math.Pow(2, Count);
            }

            return _combinationCount.Value;
        }

        public int Add(Operand operand)
        {
            int index = operand.Symbol - 'a';

            if (_operands[index] == null)
            {
                Count += 1;
                _combinationCount = null;
                _operands[index] = operand.Value;
            }

            return Count;
        }

        public void Clear()
        {
            for (int i = 0; i < OPCOUNT; i++)
            {
                _operands[i] = null;
            }
            Count = 0;
            _combinationCount = null;
        }

        public void CloneTo(Operands operands)
        {
            if (operands != this)
            {
                operands.Count = Count;
                operands._combinationCount = null;

                for (int i = 0; i < OPCOUNT; ++i)
                {
                    operands._operands[i] = _operands[i];
                }
            }
        }

        public Operands Clone()
        {
            var operands = new Operands {Count = Count};
            for (int i = 0; i < OPCOUNT; ++i)
            {
                operands._operands[i] = _operands[i];
            }

            return operands;
        }

        public int Combine(Operands operands)
        {
            if (operands != this)
            {
                for (int i = 0; i < OPCOUNT; ++i)
                {
                    if (operands._operands[i] != null && _operands[i] == null)
                    {
                        _operands[i] = operands._operands[i];
                        Count += 1;
                        _combinationCount = null;
                    }
                }
            }

            return Count;
        }

        public bool ValueOf(char c)
        {
            if (c == '1' || c == 'T') return true;
            if (c == '0' || c == 'F') return false;

            return _operands[c - 'a'] == true;
        }

        public void SetValueOf(char c, bool b)
        {
            _operands[c - 'a'] = b;
        }

        public void SetValues(int mask)
        {
            int offset = 1;
            int n = GetCombinations();

            for (int i = 0; i < OPCOUNT; ++i)
            {
                if (_operands[i] != null)
                {
                    _operands[i] = (n & (mask << offset++)) != 0;
                }
            }
        }

        public void SetValues(bool[] values)
        {
            if (Count != values.Length)
            {
                throw new ArgumentOutOfRangeException($"Expected array length of {Count}.");
            }

            int index = 0;
            for (int i = 0; i < OPCOUNT; ++i)
            {
                if (_operands[i] != null)
                {
                    _operands[i] = values[index++];
                }
            }
        }

        /// <summary>
        /// Combines operands from both arguments.
        /// </summary>
        public static Operands operator +(Operands ops1, Operands ops2)
        {
            if (ops1 == null)
            {
                return ops2?.Clone() ?? new Operands();
            }

            var cloned = ops1.Clone();

            if (ops2 == null)
            {
                return cloned;
            }

            cloned.Combine(ops2);

            return cloned;
        }

        public IEnumerator<char> GetEnumerator()
        {
            for (int i = 0; i < OPCOUNT; ++i)
            {
                if (_operands[i] != null)
                {
                    yield return (char)(i + 'a');
                }
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }
}

File: archive/src_cpp/Operands.cpp
#include "Operands.h"

namespace Logic
{

Operands::Operands()
{
	Clear();
}

Operands::Operands(const Operands& copy)
{
	this->m_iOperandCount = copy.m_iOperandCount;
	std::copy(std::begin(copy.m_acOps), std::end(copy.m_acOps), std::begin(this->m_acOps));
}

Operands::~Operands() = default;

void Operands::Clear()
{
	m_iOperandCount = 0;
	for(int i=0; i < OPCOUNT; ++i) m_acOps[i]='\0';
}

int Operands::Add(const Operand& operand)
{
	char c = operand.Symbol;
	if(c < 'a' || c > 'z') exit(1);

	if(m_acOps[CHARTOIDX(c)] == '\0')
	{
		++m_iOperandCount;
		m_acOps[CHARTOIDX(c)] = (operand.Value ? '1' : '0');
	}

	return m_iOperandCount;
}

void Operands::SetValues(unsigned long lMask)
{
	long lNbrCombos = static_cast<long>(pow(2.0, static_cast<double>(m_iOperandCount)));
	int offset=1;
	for(int i=0; i < OPCOUNT; ++i)
	{
		if(m_acOps[i] != '\0')
			m_acOps[i] = ((lNbrCombos & (lMask << offset++)) ? '1' : '0');
	}
}

void Operands::Clone(Operands& op) const
{
	if(this != &op)
	{
		op.m_iOperandCount = m_iOperandCount;
		for(int i=0; i < OPCOUNT; ++i) op.m_acOps[i] = m_acOps[i];
	}
}

int Operands::Combine(const Operands& op)
{
	if(this != &op)
	{
		for(int i=0; i < OPCOUNT; ++i)
		{
			if(op.m_acOps[i] != '\0' && m_acOps[i] == '\0')
			{
				m_acOps[i] = op.m_acOps[i];
				++m_iOperandCount;
			}
		}
	}

	return m_iOperandCount;
}

} // End namespace Logic.